server.js

```diff
@@ -7,23 +7,118 @@ const mongoUrl = process.env.MONGO_URL || "mongodb://localhost/happyThoughts"
 mongoose.connect(mongoUrl, { useNewUrlParser: true, useUnifiedTopology: true })
 mongoose.Promise = Promise
 
-// Defines the port the app will run on. Defaults to 8080, but can be 
-// overridden when starting the server. For example:
-//
+const Thought = mongoose.model('Thought', {
+  message: {
+    type: String,
+    required: true,
+    minlength: 5,
+    maxlength: 140
+  },
+  name: {
+    type: String,
+    default: "Anonymous"
+  },
+  hearts: {
+    type: Number,
+    default: 0,
+  },
+  createdAt: {
+    type: Date,
+    default: Date.now
+  }
+});
+
+// Defines the port the app will run on. 
 //   PORT=9000 npm start
-const port = process.env.PORT || 8080
-const app = express()
+const port = process.env.PORT || 8080;
+const app = express();
 
 // Add middlewares to enable cors and json body parsing
-app.use(cors())
-app.use(bodyParser.json())
+app.use(cors());
+app.use(bodyParser.json());
+
+// Middleware to handle server connection errors
+app.use((req, res, next) => {
+  try {
+    if (mongoose.connection.readyState === 1) {
+      next()
+    } else {
+      res.status(503).json({ error: 'Service unavailable' });
+    }
+  } catch (error) {
+    res.status(400).json({ error: 'Error! Could not access the server.' });
+  }
+});
+
+////////////// API ENDPOINTS ////////////// 
 
-// Start defining your routes here
+// To list all available endpoints on the starting page
+const listEndpoints = require('express-list-endpoints');
 app.get('/', (req, res) => {
-  res.send('Hello world')
-})
+  res.send(listEndpoints(app));
+});
+
+// Endpoint created to get a list of the 20 latest thoughts
+app.get('/thoughts', async (req, res) => {
+  // Field for client to sort on parameter
+  const sortField = req.query.sortField;
+  // Field for client to choose sorting order, 'asc' or 'desc'
+  const sortOrder = req.query.sortOrder;
+  // Creates the initial database query
+  let databaseQuery = Thought.find();
+
+  // To search with sortField and sortOrder with queries: 
+  // (GET '/thoughts?sortField=${sortField}&sortOrder=${sortOrder}')
+  if (sortField) {
+    databaseQuery = databaseQuery.sort({ [sortField]: sortOrder === 'asc' ? 1 : -1 })
+  } else {
+    databaseQuery = databaseQuery.sort({ createdAt: 'desc' }).limit(20);
+  }
```

Comment:
When you created so nice functionality in your API why didn't you use it in your frontend?

Reply:
I didn't have time but I will really try to now! 😊